strawberry/fastapi/__init__.py

```diff
@@ -0,0 +1,188 @@
+import json
+from datetime import timedelta
+from typing import Any, Optional
+
+from fastapi import (
+    APIRouter,
+    BackgroundTasks,
+    Depends,
+    Request,
+    Response,
+    WebSocket,
+    status,
+)
+from fastapi.responses import HTMLResponse, JSONResponse, PlainTextResponse
+from strawberry.asgi.utils import get_graphiql_html
+from strawberry.exceptions import MissingQueryError
+from strawberry.fastapi.handlers.graphql_transport_ws_handler import (
+    GraphQLTransportWSHandler,
+)
+from strawberry.fastapi.handlers.graphql_ws_handler import GraphQLWSHandler
+from strawberry.file_uploads.utils import replace_placeholders_with_files
+from strawberry.http import GraphQLHTTPResponse, parse_request_data, process_result
+from strawberry.schema import BaseSchema
+from strawberry.subscriptions import GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL
+from strawberry.types import ExecutionResult
+from strawberry.utils.debug import pretty_print_graphql_operation
+
+
+async def get_root_value():
+    return None
+
+
+async def get_context(
+    background_tasks: BackgroundTasks, request: Request = None, ws: WebSocket = None
+) -> Optional[Any]:
+    return {"request": request or ws, "background_tasks": background_tasks}
+
+
+class GraphQL(APIRouter):
+    graphql_ws_handler_class = GraphQLWSHandler
+    graphql_transport_ws_handler_class = GraphQLTransportWSHandler
+
+    def __init__(
+        self,
+        schema: BaseSchema,
+        prefix: str = "",
+        graphiql: bool = True,
+        keep_alive: bool = False,
+        keep_alive_interval: float = 1,
+        debug: bool = False,
+        root_value_getter=get_root_value,
+        context_getter=get_context,
+        subscription_protocols=(GRAPHQL_TRANSPORT_WS_PROTOCOL, GRAPHQL_WS_PROTOCOL),
+        connection_init_wait_timeout: timedelta = timedelta(minutes=1),
+    ):
+        super().__init__(prefix=prefix)
+        self.schema = schema
+        self.graphiql = graphiql
+        self.keep_alive = keep_alive
+        self.keep_alive_interval = keep_alive_interval
+        self.debug = debug
+        self.root_value_getter = root_value_getter
+        self.context_getter = context_getter
+        self.protocols = subscription_protocols
+        self.connection_init_wait_timeout = connection_init_wait_timeout
+
+        @self.get(
```

Comment:
There's no need to register this endpoint if the `self.graphiql` is `False` to begin with, and you don't need to add the logic around 404, as it will be automatically handled.

There's no need for `response_class` parameter if you are returning it anyway.
```python
        if self.graphiql:
            self.add_api_route(
                "",
                self.get_graphiql_response,
                methods=["GET"],
                responses={
                    200: {
                        "description": "The GraphiQL integrated development environment.",
                    },
                },
            )
```
I think `add_api_route` is cuter on cases like this, but you can also use the `get` decorator.

Reply:
If the endpoint doesn't work because of the `self`, you can turn it into a `classmethod`.

Reply:
GraphiQL is registered on same endpoint as HTTP and websocket handlers, so GET request would responded with 405 - method not allowed. ASGI, Flask and other extensions are using 404 status so to keep it consistent I'm using 404 too, although I think 405 would be more appropriate.

Reply:
I agree that `add_api_route` is better in this case.